File: High Frequency/SimStudyPreav.R
nsim = 10000
n = 23400 # 1 sec sampling in 6.5 trading hours
theta = 1 # tuning parameter of pre-averaging

sigma0 = 0.04/250
kappa = 5/250
xi = 0.5/250
rho = -0.5
gamma = 0.5 # Noise to signal ratio

P = array(NA, dim = nsim)
IV = array(NA, dim = nsim)

for (i in 1:nsim){
  print(i)
  
  heston = f_SVHeston(1,n,sigma0, kappa,xi,rho)
  X = heston[["X"]]
  sigma = heston[["sigma"]]
  IV[i] = mean(sigma[1:(length(sigma)-1)]**2)
  
  
  ## Adding noise to prices
  omega = gamma*sqrt(IV[i]/n) #omega**2 is variance of noise
  noise = omega*rnorm((n+1)) # add iid gaussian npise
  Y = X + noise
  
  # Pre-averaging
  
  K = round(theta*sqrt(n))
  P[i] = f_preav(Y,K)
  
  out = list()
  
  out[["P"]] = P 
  out[["IV"]] = IV
}

K

bias = mean(P-IV)
bias

Rbias = mean(P/IV-1)
Rbias

MSE = mean((P-IV)**2)
MSE

1.866552977303856e-10

IV
P

File: High Frequency/Matlab/SimStudyPreav.m
%--------------------------------------------------------------------------
%  This code computes the bias and MSE of the pre-averaging estimator 
% with noisy high-frequency data and n=23400 returns, i.e. 1s sampling.
%--------------------------------------------------------------------------

rng('default'); %uses default random number generator. 
%Results (bias, MSE) depend slighlty on the random number generator
%Setting parameters
nsim = 10000;  %number of simulations
n = 23400;  % i.e. 1 sec sampling in 6.5 trading hours
theta = 1;  %tuning parameter of pre-averaging,
[sigma0,kappa,xi,rho] = deal(0.04/250,5/250,0.50/250,-0.50);
gamma = 0.5; % noise to signal ratio
%Simulation loop starts and P denotes the pre-averaging estimator
[P, IV] = deal(NaN(nsim,1));
for i = 1 : nsim
    fprintf('Sim no. %5d of %5d...\n',i,nsim); tic;
    [X,sigma] = f_SVHeston(1, n,sigma0,kappa,xi,rho); %Simulate Heston 
    IV(i) = mean(sigma(1:end-1).^2); %IV is approximated by a sum
     %Adding noise to prices
      omega = gamma*sqrt(IV(i)/n); %omega^2 is variance of the noise
      noise = omega*randn(n+1,1); %add iid gaussian noise
      Y =X+noise; %Y is the observed price
      % pre-averaging
      K= round(theta*sqrt(n)); %block size k_n
      P(i) = f_preav(Y,K);
      fprintf('Time elapsed: %5.2f...\n',toc);
end
bias=mean(P-IV);
Rbias=mean(P./IV-1);
MSE=mean((P-IV).^2);